pkg/sql/prepared_stmt.go

```diff
@@ -127,16 +123,22 @@ type PreparedPortal struct {
 	// OutFormats contains the requested formats for the output columns.
 	OutFormats []pgwirebase.FormatCode
 
-	// refCount keeps track of the number of references to this PreparedStatement.
-	// New references are registered through incRef().
-	// Most references are being held by portals created from this prepared
-	// statement.
-	refCount int
-
 	// exhausted tracks whether this portal has already been fully exhausted,
 	// meaning that any additional attempts to execute it should return no
 	// rows.
 	exhausted bool
+
+	// mutablePreparedPortalState keeps track of the reference counter. This
+	// needs to be mutable and is, thus, shared among all references.
+	*mutablePreparedPortalState
+}
+
+type mutablePreparedPortalState struct {
```

Comment:
thanks for explaining in the PR! that makes me wonder... maybe we could just keep using the `PreparedPortal` by value (and somehow require that it's only used by value). it seems like we only need to care about counting references to `Stmt  *PreparedStatement`. so would things work if we deleted `mutablePreparedPortalState` entirely and made this update instead?

```
-// decRef decrements the number of references to this portal. If the refCount
-// reaches 0, then the memory account is shrunk accordingly.
+// decRef decrements the number of references to the prepared statement used
+// by this portal. The memory account is shrunk accordingly.
 func (p *PreparedPortal) decRef(
     ctx context.Context, prepStmtsNamespaceMemAcc *mon.BoundAccount, portalName string,
 ) {
-    if p.refCount <= 0 {
-        log.Fatal(ctx, "corrupt PreparedPortal refcount")
-    }
-    p.refCount--
-
-    if p.refCount == 0 {
-        prepStmtsNamespaceMemAcc.Shrink(ctx, p.size(portalName))
-        p.Stmt.decRef(ctx)
-    }
+    prepStmtsNamespaceMemAcc.Shrink(ctx, p.size(portalName))
+    p.Stmt.decRef(ctx)
 }
```